### PR TITLE
[v5.0.1] Add threshold for maxTreeDepth for serializers

### LINKED ISSUE #759: Implement all TODO v5.x
This is tracking issue for implementation of all TODO items in source code marked with 
v5.x, for example `// TODO v5.x (3h): this method must be removed`.

### PR DESCRIPTION
**Pull request by jozanek**
Partially closes #759, #781

### CODE REVIEW

File: sigmastate/src/test/scala/sigmastate/serialization/DeserializationResilience.scala [hunk 1]
```diff
@@ -255,6 +288,12 @@ class DeserializationResilience extends SerializationSpecification
       ValueSerializer.deserialize(reader(ValueSerializer.serialize(expr), maxTreeDepth = 3))
```
Comment (aslesarenko) [inline sigmastate/src/test/scala/sigmastate/serialization/DeserializationResilience.scala:290]: Please make sure the thrown exception comes from TypeSerializer. I.e. that the test is make sense.

File: sigmastate/src/test/scala/sigmastate/serialization/DeserializationResilience.scala [hunk 2]
```diff
@@ -255,6 +288,12 @@ class DeserializationResilience extends SerializationSpecification
...
   }
 
+  property("max recursive call depth is checked in writer.level for TypeSerializer") {
+    val expr = Tuple(Tuple(IntConstant(1)))
+    an[SerializeCallDepthExceeded] should be thrownBy
+      ValueSerializer.serialize(expr, writer(maxTreeDepth = 2))
```
Comment (aslesarenko) [inline sigmastate/src/test/scala/sigmastate/serialization/DeserializationResilience.scala:296]: Same issues as with the previous test I think.

**Review by aslesarenko — CHANGES_REQUESTED** [review]

File: sigmastate/src/test/scala/sigmastate/serialization/DeserializationResilience.scala (hunk 1 above)
Comment (jozanek) [reply sigmastate/src/test/scala/sigmastate/serialization/DeserializationResilience.scala:377]: Updated in new version.

File: sigmastate/src/test/scala/sigmastate/serialization/DeserializationResilience.scala (hunk 2 above)
Comment (jozanek) [reply sigmastate/src/test/scala/sigmastate/serialization/DeserializationResilience.scala]: Correct, the TypeSerializer was completely missing handling for maxTreeDepth for both de/serializing, added to both method and updated tests.

**Review by aslesarenko — CHANGES_REQUESTED** [review]

File: sigmastate/src/main/scala/sigmastate/serialization/TypeSerializer.scala [hunk 3]
```diff
@@ -23,100 +23,105 @@ object TypeSerializer {
...
+          serializeTuple(tup, w)
+      }
+      case typeIdent: STypeVar => {
+        w.put(typeIdent.typeCode)
+        val bytes = typeIdent.name.getBytes(StandardCharsets.UTF_8)
+        w.putUByte(bytes.length)
+          .putBytes(bytes)
+      }
     }
+    w.level = depth
```
[the remark this answers is not in the document]
Comment (jozanek) [reply sigmastate/src/main/scala/sigmastate/serialization/TypeSerializer.scala:119]: Github diff tool is quite puzzled, had to break formatting to show it properly. I can add it in additional commit after PR is reviewed.

File: sigmastate/src/main/scala/sigmastate/serialization/TypeSerializer.scala [hunk 4]
```diff
@@ -199,12 +204,13 @@ object TypeSerializer {
...
     }
+    r.level = depth
     tpe
   }
 
   private def getArgType(r: SigmaByteReader, primId: Int, depth: Int) =
     if (primId == 0)
-      deserialize(r, depth + 1)
+      deserialize(r)
     else
```
[the remark this answers is not in the document]
Comment (jozanek) [reply sigmastate/src/main/scala/sigmastate/serialization/TypeSerializer.scala]: done

**Comment by jozanek** [comment]
@aslesarenko can you review again please and possibly merge?

**Comment by aslesarenko** [comment]
@jozanek, this is low prioriity and considering how much unmerged code is there in `develop` already I'm putting this PR on hold.

**Comment by aslesarenko** [comment]
After giving more thought into this I think this feature will complicate maintenance.
Depth limits on serialisation should exactly match measurements on deserialisation. 
Any difference may lead to subtle bugs in dApps using Sigma (Appkit, Fleet).
And adversary can create malicious bytes directly anyway, so the value of this checks is small.

**Pull request closed without merging**